### ISSUE TITLE
xfdashboard takes significant CPU when a window moves/changes

### ISSUE THREAD

**Issue by alejandro-perez** [gmc-holle/xfdashboard#44]
xfdashboard takes significant when a window changes its position, size. This can be most of the time ok, but it happens all the time during video playback. Is this the expected behaviour?

Regards

**Comment by alejandro-perez** [comment]
I'd like to re-open this issue. While watching movies is not a problem anymore, moving a window is. Test to reproduce:

1) Open a terminal, execute top to see xfdashboard
2) Open a large window, such as web browser, thunar... Make it as big as possible.
3) Drag the window and, whitout undragging it, move the window around. 
4) While moving the window, see how xfdashboard gets a high CPU (50% in my case).

While this situation is not noticeable in my daily basis, I noticed it at work, where I have two screens and move big windows from one to the other quite often. The movement is far from being smooth due to xfdashboard. 

I guess it might be related to the window_tracking functionality. Could this be avoided during drag operations, and just be called after the window has been finally moved?

Regards

**Comment by gmc-holle** [comment]
I see. I have two solutions for this issue:

1.) The simpler solution would be not to emit the signal 'window-geometry-changed' that a window's position or size changed at all when application is suspended (hidden). Once the application is resumed (shown again) I iterate through all windows and emit this signal for then regardless if their position or size have changed in the meantime. That would cause a huge load for a short time because a lot of actors are listening to this signal and their signal handlers may be called for each window.

2.) The more complex solution would be to introduce a timeout source with a e.g. 100ms timeout when the application is suspended. The timer of this source is resettet if it didn't reach the timeout and the window is moved or resize in the meantime. Once the timeout is reached the signal is emitted. This would cause lower cpu load but is more complex and complex solution could fail easier :/

I would prefer the first solution. What do you think?

**Comment by alejandro-perez** [comment]
I think 1 is ok, as long as it does not need more time to "recover" than just running xfdashboard (no daemon). Otherwise, I would just not run xfdashboard in background and call it whenever I need.

The idea I had in mind is that you just send the signal when the window is dropped, not while it's being dragged. I though you were able to tell the difference :(. Ej. in xfwm4 you are able to hide the content while moving, so I thought you got somehow a signal of window dropped.

You could create a branch (to avoid making things dirty) and try option 1. In any case, in lower-end computers, if moving a window supposes me a 50% CPU, I'd rather have it once, at xfdashboard call :).

**Comment by gmc-holle** [comment]
Option 1 should be faster than starting up xfdashboard because it does only need to handle these signal emissions which mostly call a relayout of actors and which - in turn - shoud be queued to one relayout. So I think I go this way ;)

**Comment by alejandro-perez** [comment]
That would be superb. Thanks!

**Comment by gmc-holle** [comment]
It's getting better I think ;) Combined with the idea in #88 I should be able to reduce the signal emissions only to these windows which have moved or resized in the meantime resulting in lower cpu load.

**Comment by gmc-holle** [comment]
I hope it is fixed with commit b9bccc9079d294bee1f251a57024dcf811bcc789

**Comment by alejandro-perez** [comment]
Indeed it does. And quite fast I would say :). Thanks!

**Comment by gmc-holle** [comment]
Thanks for reporting :D

**Comment by alejandro-perez** [comment]
It seems that this patch is generating more problems than the one I commented before. For a time I have notice that when I came back from suspension, thunderbird get frozen 50% of the times. I have temporarily disabled xfdashboard and it seems to work again. Do it make any sense, or may be caused by a different thing?

**Comment by gmc-holle** [comment]
I don't say it is not possible but I doubt it. I don't think that this application can cause other application to hang or to lock up.

Indeed, the patch had a bug because it connected a new signal handler for window movement and resizes next to the existing blocked ones each time the daemon was resumed. This should be fixed now.

Could you reapply the first patch fo commit b9bccc9079d294bee1f251a57024dcf811bcc789 and also the new one at commit dd9039ad352df3e2d77b0550d5e2577b1287871d and test it?

**Comment by alejandro-perez** [comment]
I would... if I knew how to do it. I'm not well versed in git. Would it work if I just get the latest version?

**Comment by gmc-holle** [comment]
If you had run "git revert ..." before then your local git repository has changed - that's how I understand git.

I would try this commands in that case:

```
git stash
(git stash drop)
git pull
```

But maybe just a ``git pull``` is enough. I'm not a git pro also ;)

**Comment by alejandro-perez** [comment]
Well, I'm dumber, I would say. Since I don't like to have "manually installed" stuff, every time I get the xfdashboard-git package from AUR, build it, and install the package. I don't really have the local repository :). So everytime is a clean build. I will do it right away.

**Comment by gmc-holle** [comment]
Then it should "just-work"(tm)  ;)

**Comment by alejandro-perez** [comment]
Nevertheless, I just recalled I cannot directly test it here, since the major problem was at work. I will check if it interferes with Thunderbird.

I'm not saying that it really messes Thunderbird, but it might get the WM to an inconsistent state, where the content of the TB window is not drawn. In any case, I will test it!

**Comment by gmc-holle** [comment]
Thanks

**Comment by alejandro-perez** [comment]
Hi,

the problem with TB is gone, so +1. Tomorrow I will test it at work.

Regards

**Comment by alejandro-perez** [comment]
Hi, I tested it at work, and it is still pretty unstable. I'm not sure of the cause, but half of the times I call "xfdashboard -t" here it hangs and behaves weird. 

Besides, I have to regret of my previous comment, as this morning TB was dead (still not sure if related).

**Comment by alejandro-perez** [comment]
More info. The TB problem seems to not be related to xfdashboard. So it is only the stability problem at work.

Regards

**Comment by gmc-holle** [comment]
I would have been very surprised if the last commit would fix your problem with TB suddenly because I cannot think how it would affect it. But the commit fixes a memory leak by adding and adding new signal handlers over and over again.

I also faced a stability issue with xfdashboard running in daemon mode now. It happens after running for some days non-stop. Maybe issue #49 (I just closed some days ago) is valid? I need to run the daemon mode in gdb to catch the cause. I only see in my .xsession-error log file that the segfault happens in GDK3 library.

**Comment by alejandro-perez** [comment]
Hi,

anyway, the problem in my computer at work is far worse than that. I 
cannot run it without eating the whole CPU even in foreground mode.

**Comment by gmc-holle** [comment]
Does it eat your CPU even if it suspended and you are do nothing even not moving the mouse? Or is CPU usage only high if you move or resize a window?

**Comment by alejandro-perez** [comment]
The problem at work is worse. When I start xfdashboard (no daemon), half 
of the times it freezes the computer for like 30s, and the CPU get's 
100% all the time.

Sometimes I have to SSH from the smartphone and kill xfdashboard. I 
think this was caused by one of the fixes for this Issue. But since I'm 
using Arch (rolling release) it might be caused by many other reasons 
(change on kernel, graphic driver...). I will try to revert this patch 
tomorrow and see if it is solved.

**Comment by gmc-holle** [comment]
If it doesn't work could you revert to version 0.4? Just to check if it
is more stable and only eats CPU when moving or resizing windows.


Am Donnerstag, den 14.05.2015, 11:28 -0700 schrieb Alejandro Perez:

**Comment by alejandro-perez** [comment]
Hi!

I can finally "blame" a commit: 729bbb9ff2e012ecf995691bb7bd9e510cae0e68. I have download xfdashboard and reverted commit by comment until I found out which one was the cause (I think git bisect was created for that purpose...but I did it the manual way:D). 

To further confirm I the re-downloaded a clean copy and reverted just that specific commit. Everything works now (without the functionality of that commit, of course).

Hope this helps!

**Comment by gmc-holle** [comment]
Now it is getting weird. I can reproduce a crash of xfdashboard running in daemon mode by just creating a new window and closing it while the application is suspended but was resumed at least once in the meantime. :(

Steps to reproduce:

* Start xfdashboard in daemon mode
* Resume xfdashboard
* Suspend xfdashboard
* Open a window (e.g. start a terminal)
* Close this new window
* Enjoy the crash :(

This is the stacktrace as shown in gdb:

```
Program received signal SIGSEGV, Segmentation fault.
0x0000000003ebb2d0 in ?? ()
(gdb) bt
#0  0x0000000003ebb2d0 in ?? ()
#1  0x00007ffff6b47577 in _cogl_winsys_texture_pixmap_x11_free () from /usr/lib64/libcogl.so.20
#2  0x00007ffff6b4563b in _cogl_object_texture_pixmap_x11_indirect_free () from /usr/lib64/libcogl.so.20
#3  0x000000000047b78c in _xfdashboard_window_content_release_resources (self=0x3f6dc30) at window-content.c:512
#4  0x000000000047b8dc in _xfdashboard_window_content_dispose (inObject=0x3f6dc30) at window-content.c:1379
#5  0x00007ffff4d2fff3 in g_object_unref () from /usr/lib64/libgobject-2.0.so.0
#6  0x00007ffff6fdfceb in clutter_actor_dispose () from /usr/lib64/libclutter-1.0.so.0
#7  0x00007ffff4d31da8 in g_object_run_dispose () from /usr/lib64/libgobject-2.0.so.0
#8  0x00007ffff6fd384e in clutter_actor_destroy () from /usr/lib64/libclutter-1.0.so.0
#9  0x00007ffff4d2b620 in g_closure_invoke () from /usr/lib64/libgobject-2.0.so.0
#10 0x00007ffff4d3d215 in signal_emit_unlocked_R () from /usr/lib64/libgobject-2.0.so.0
#11 0x00007ffff4d44fba in g_signal_emit_valist () from /usr/lib64/libgobject-2.0.so.0
#12 0x00007ffff4d45262 in g_signal_emit () from /usr/lib64/libgobject-2.0.so.0
#13 0x00007ffff4d2e837 in g_cclosure_marshal_VOID__OBJECTv () from /usr/lib64/libgobject-2.0.so.0
#14 0x00007ffff4d2b847 in _g_closure_invoke_va () from /usr/lib64/libgobject-2.0.so.0
#15 0x00007ffff4d445f7 in g_signal_emit_valist () from /usr/lib64/libgobject-2.0.so.0
#16 0x00007ffff4d45262 in g_signal_emit () from /usr/lib64/libgobject-2.0.so.0
#17 0x00007ffff7bb2015 in do_update_now () from /usr/lib64/libwnck-3.so.0
#18 0x00007ffff7bb2974 in update_idle () from /usr/lib64/libwnck-3.so.0
#19 0x00007ffff4a2b376 in g_main_context_dispatch () from /usr/lib64/libglib-2.0.so.0
#20 0x00007ffff4a2b6c8 in g_main_context_iterate.isra () from /usr/lib64/libglib-2.0.so.0
#21 0x00007ffff4a2b9f2 in g_main_loop_run () from /usr/lib64/libglib-2.0.so.0
#22 0x00007ffff701560b in clutter_main () from /usr/lib64/libclutter-1.0.so.0
#23 0x000000000041fc5c in main (argc=2, argv=0x7fffffffdd58) at main.c:293
```

Once I also noticed that the system was unusable (but **no** CPU core was at 100%) and I had to kill xfdashboard. Maybe it is related to the same source causing the crash.

After reverting to commit 9c527f22b9c12fc7875efb64d6dedbf03f4e17d7, the one before the commit 729bbb9ff2e012ecf995691bb7bd9e510cae0e68, the problem is gone. So I agree that commit 729bbb9ff2e012ecf995691bb7bd9e510cae0e68 is to blame.

**Comment by alejandro-perez** [comment]
Well, I didn't need to create a window and then delete it to get the system unusable, so I guess it should be something related to memory or pointers pointing who knows where.

In any case, in my laptop it has been working well with the latests version, so I didn't revert anything. So I have tried your sequence, but it didn't work (I mean, it didn't fail). I guess is random, which are the unfunniest kind :D

**Comment by alejandro-perez** [comment]
Just to be clear, I cannot make it fail in my laptop, not foloowing your instructions, but I can make it fail in my work computer by just starting it. After reverting, it works quite fine at work.

**Comment by gmc-holle** [comment]
At home it doesn't fail either. So I guess it might be related to the Nvidia graphics driver because that's the significant difference between my two machines. The one at home uses the in-cpu Intel GPU while at work it is Nvidia. If I remember right that's also the difference between your machines, isn't it?

Nevertheless there's a bug but it behaves slightly different at different computers. That will be fun to track down the source of the problem. If I should guess I would say it looks like a use-after-free pointer issue.

**Comment by alejandro-perez** [comment]
That's right, I have a NVIDIA at work, and an Intel at home. At work I have two monitors, if that matters.

For finding user-after-free errors there are several libraries we can use (e.g. electric fence), or even valgrind. Have you tried them? I think the first one is as easy as linking your program with -lefence, whereas the last one is a complete emulator.

BTW, I resolved the TB problem, that wasn't related to xfdashboard at all. It was a bug with firetray (extension) and libappindicator-gtk2.

Cheers!

**Comment by alejandro-perez** [comment]
Hi again,

I made a quick test at home (where it does not fail), and efence didn't report anything. I should try on Monday at work. valgrind didn't report any use-after-delete error. But, there are many memory leak reports. 

Hope this helps. I guess the only ones relevant are the ones marked as "definitely lost". 

https://www.dropbox.com/s/1cwiprg5v6fumdl/leaks.txt?dl=0

Regards

**Comment by gmc-holle** [comment]
Thanks for providing the valgrind log file. I tried to locate and to solve most of the claimed locations. I also did run my own valgrind test and found a lot of unfreed and unreferenced variables. Most of them should be solved now although I believe there are many more mem-leaks left.

Although I do not believe that it will change the origin of the last issue at this problem I ask you to test again for this problem. I cannot reproduce this problem at home with an Intel GPU and a decidicated Nvidia card. And at work I'm very very busy at the moment so I can only confirm problems but not solve them.

**Comment by alejandro-perez** [comment]
Surely fixing memory leaks should not solve any problem, but maybe one of them, in addition to a leak, was also the source of the problem. Anyhow, I will test it at work (where it fails always) with efence and valgrind, and see if I can bring anything to ligth. 

Thanks for the hard work and great tool!

**Comment by alejandro-perez** [comment]
Well, sadly I couldn't . At work I get it to fail quite easily (I just need to go through the different workspaces twice and I got it hanged, 100% CPU). But.... 100% CPU of Xorg processes, not xfdashboard one.

Hope that helps.

**Comment by gmc-holle** [comment]
It's getting more worse here too when using the system running the binary Nvidia driver. I also sometime have problems suspending the application, i.e. the fullscreen window will not disappear anymore.

For now I will revert commit 729bbb9ff2e012ecf995691bb7bd9e510cae0e68. This also means you will not be able to theme the windows in the small workspaces anymore. 

Closing this issue because the initial problem reported here is solved but reopening #89 because of this issue this problematic commit was created.

**Comment by alejandro-perez** [comment]
Yes, I do use the binary driver too. Without commit 729bbb9 it works like a charm (so far). I think it makes sense to revert that commit. Moreover, I'm not using that theme as much as I though, since I do usually like to see the preview and my boss is not always behind my back :). So, it would be totally ok if you want to focus on any other issue and leave #89 for the future.

Thanks!
